dodiscover/context_builder.py

```diff
@@ -0,0 +1,140 @@
+from typing import Optional, Set, Union
+
+import networkx as nx
+import pandas as pd
+
+from ._protocol import Graph
+from .context import Context
+
+
+class ContextBuilder:
+    """A builder class for creating Context objects ergonomically.
+
+    The context builder provides a way to capture assumptions, domain knowledge, and data.
+
+    Parameters
+    ----------
+    data : pd.DataFrame
+        A dataset in data-frame form, consisting of samples as rows and variables as columns.
+    """
+
+    _graph: Optional[Graph] = None
+    _included_edges: Optional[Union[nx.Graph, nx.DiGraph]] = None
+    _excluded_edges: Optional[Union[nx.Graph, nx.DiGraph]] = None
+    _observed_variables: Optional[Set[str]] = None
+    _latent_variables: Optional[Set[str]] = None
+
+    def __init__(self, data: pd.DataFrame) -> None:
+        self._data = data
+
+    def data(self, data: pd.DataFrame) -> "ContextBuilder":
+        """Set the dataset to use.
+
+        Parameters
+        ----------
+        data : pd.DataFrame - the new dataframe to use
+
+        Returns
+        -------
+        ContextBuilder
+            The builder instance
+        """
+        self._data = data
+        return self
+
+    def graph(self, graph: Graph) -> "ContextBuilder":
+        """Set the partial graph to start with.
+
+        Parameters
+        ----------
+        data : pd.DataFrame - the new dataframe to use
+
+
+        Returns
+        -------
+        ContextBuilder
+            The builder instance
+        """
+        self._graph = graph
+        return self
+
+    def edge_constraints(
+        self,
+        included_edges: Optional[Union[nx.Graph, nx.DiGraph]] = None,
+        excluded_edges: Optional[Union[nx.Graph, nx.DiGraph]] = None,
+    ) -> "ContextBuilder":
+        """Set edge constraints to apply in discovery.
+
+        Parameters
+        ----------
+        included_edges : Optional[Union[nx.Graph, nx.DiGraph]]
+            Edges that should be included in the resultant graph
+        excluded_edges : Optional[Union[nx.Graph, nx.DiGraph]]
+            Edges that must be excluded in the resultant graph
+
+        Returns
+        -------
+        ContextBuilder
+            The builder instance
+        """
+        self._included_edges = included_edges
+        self._excluded_edges = excluded_edges
+        return self
+
+    def features(
```

Comment:
I think we should say "variables."  This library will also take on some causal representation learning, and in that setting causal variables and "features" are different (features will be a lower level of abstraction.

Reply:
If we need something in the constructor, I vote for this